lightstep/sdk/metric/aggregator/sum/sum.go

```diff
@@ -0,0 +1,154 @@
+	return &NonMonotonicInt64{value: x}
+}
+
+func NewMonotonicFloat64(x float64) *MonotonicFloat64 {
+	return &MonotonicFloat64{value: x}
+}
+
+func NewNonMonotonicFloat64(x float64) *NonMonotonicFloat64 {
+	return &NonMonotonicFloat64{value: x}
+}
+
+func (Monotonic) kind() aggregation.Kind {
+	return aggregation.MonotonicSumKind
+}
+
+func (NonMonotonic) kind() aggregation.Kind {
+	return aggregation.NonMonotonicSumKind
+}
+
+func (Monotonic) category() aggregation.Category {
+	return aggregation.MonotonicSumCategory
+}
+
+func (NonMonotonic) category() aggregation.Category {
+	return aggregation.NonMonotonicSumCategory
+}
+
+var (
+	_ aggregator.Methods[int64, MonotonicInt64]        = Methods[int64, number.Int64Traits, Monotonic, MonotonicInt64]{}
+	_ aggregator.Methods[float64, MonotonicFloat64]    = Methods[float64, number.Float64Traits, Monotonic, MonotonicFloat64]{}
+	_ aggregator.Methods[int64, NonMonotonicInt64]     = Methods[int64, number.Int64Traits, NonMonotonic, NonMonotonicInt64]{}
+	_ aggregator.Methods[float64, NonMonotonicFloat64] = Methods[float64, number.Float64Traits, NonMonotonic, NonMonotonicFloat64]{}
+
+	_ aggregation.Sum = &MonotonicInt64{}
+	_ aggregation.Sum = &MonotonicFloat64{}
+	_ aggregation.Sum = &NonMonotonicInt64{}
+	_ aggregation.Sum = &NonMonotonicFloat64{}
+)
+
+func (s *State[N, Traits, M]) Sum() number.Number {
+	var t Traits
+	return t.ToNumber(s.value)
+}
+
+func (s *State[N, Traits, M]) Kind() aggregation.Kind {
+	var m M
+	return m.kind()
+}
+
+func (s *State[N, Traits, M]) IsMonotonic() bool {
+	var m M
+	return m.kind() == aggregation.MonotonicSumKind
+}
+
+func (Methods[N, Traits, M, Storage]) Kind() aggregation.Kind {
+	var m M
+	return m.kind()
+}
+
+func (Methods[N, Traits, M, Storage]) Init(state *State[N, Traits, M], _ aggregator.Config) {
+	// Note: storage is zero to start
+}
+
+func (Methods[N, Traits, M, Storage]) Move(src, dest *State[N, Traits, M]) {
+	var t Traits
+	dest.value = t.SwapAtomic(&src.value, 0)
+}
+
+func (Methods[N, Traits, M, Storage]) HasChange(ptr *State[N, Traits, M]) bool {
+	return ptr.value != 0
+}
+
+func (Methods[N, Traits, M, Storage]) Update(state *State[N, Traits, M], value N) {
+	var t Traits
+	t.AddAtomic(&state.value, value)
+}
+
+func (Methods[N, Traits, M, Storage]) Copy(from, to *State[N, Traits, M]) {
+	var t Traits
+	to.value = t.GetAtomic(&from.value)
+}
+
+func (Methods[N, Traits, M, Storage]) Merge(from, to *State[N, Traits, M]) {
+	var t Traits
+	t.AddAtomic(&to.value, from.value)
+}
+
+func (Methods[N, Traits, M, Storage]) ToAggregation(state *State[N, Traits, M]) aggregation.Aggregation {
+	return state
+}
+
+func (Methods[N, Traits, M, Storage]) ToStorage(aggr aggregation.Aggregation) (*State[N, Traits, M], bool) {
+	r, ok := aggr.(*State[N, Traits, M])
+	return r, ok
+}
+
+func (Methods[N, Traits, M, Storage]) SubtractSwap(newvalue, oldvalueModified *State[N, Traits, M]) {
+	oldvalueModified.value = newvalue.value - oldvalueModified.value
```

Comment:
Shouldn't this also be using the atomics methods from traits?

Reply:
Good question. This deserves more comments. Since this is used only for asynchronous instruments, the answer is no, but that's not guaranteed by anything. I can't see any reason to use Subtract for synchronous instruments, where it would matter, unless we start talking about synchronous Gauges. If we did, we'd want to have a consistent value to subtract, so would want to read the variable via Copy or Move the way the synchronous-stateless collector does. This leads again, to no, no synchronization is needed for subtraction.